File: java_solutions_1/454.4sum-ii.java
/*
 * @lc app=leetcode id=454 lang=java
 *
 * [454] 4Sum II
 */

class Solution {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        return java.util.stream.Stream.of(new java.util.HashMap<Integer, Integer>()).peek(map -> java.util.Arrays.stream(nums1).forEach(u -> java.util.Arrays.stream(nums2).forEach(v -> map.put(u + v, map.getOrDefault(u + v, 0) + 1)))).map(map -> java.util.Arrays.stream(nums3).flatMap(u -> java.util.Arrays.stream(nums4).map(v -> map.getOrDefault(-(u + v), 0))).sum()).findFirst().get();
    }
}

File: kotlin_solutions/454.4-sum-ii.kt
/*
 * @lc app=leetcode id=454 lang=kotlin
 *
 * [454] 4Sum II
 */

class Solution { fun fourSumCount(nums1: IntArray, nums2: IntArray, nums3: IntArray, nums4: IntArray): Int = nums1.flatMap { a -> nums2.map { b -> a + b } }.groupingBy { it }.eachCount() .let { map -> nums3.flatMap { c -> nums4.map { d -> c + d } }.sumOf { map.getOrDefault(-it, 0) } } }